### PR TITLE
METRON-2169: Upgrade Kafka/Storm

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This PR upgrades Kafka 2.0.0 and Storm to 1.2.0.

### Changes Included

- Updated Kafka and Storm versions in Maven
- Updated Maven pom files to include the correct dependency versions
- Updated Maven profile in Ansible to HDP-3.1.0.0
- Changed REST classes to use new Kafka client classes
- Substituted Kafka constants for hardcoded strings
- Upgraded to the `KafkaComponent` class to use new Kafka classes

### Testing
I have done some light testing in full dev including:

- Verified all topologies are running with no errors and data is flowing through
- Verified the Kafka REST endpoints functions as expected
- Verified the Kafka Stellar functions continue to work

### Future Work
I believe there are parts of our codebase that we can deprecate in separate PRs:

- https://issues.apache.org/jira/browse/METRON-2170
- https://issues.apache.org/jira/browse/METRON-2171

I also ran into several classpath version errors.  It is not possible to resolve these using the standard Maven exclude tags due to the issue described [here](https://github.com/apache/metron/pull/1436).  As a workaround, I move the conflicting dependency to the top of the pom so that it takes priority over the shaded dependencies.  This is not ideal but easier than trying to track the version down in the shaded dependencies.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

- [ ] Have you ensured that any documentation diagrams have been updated, along with their source files, using [draw.io](https://www.draw.io/)? See [Metron Development Guidelines](https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines) for instructions.

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -141,6 +141,17 @@
...
 
     <profiles>
+        <profile>
+            <id>HDP-3.1.0.0</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
+            </activation>
+            <properties>
+                <hdp_version>3.1.0.0</hdp_version>
+                <global_storm_version>1.2.1</global_storm_version>
```
Comment (mmiklavc) [inline pom.xml]: Can we separate Storm and Kafka? These are not closely related the way ES and Kibana are.

File: metron-platform/metron-integration-test/pom.xml [hunk 2]
```diff
@@ -130,7 +130,7 @@
     </dependency>
     <dependency>
       <groupId>org.apache.kafka</groupId>
-      <artifactId>kafka_2.10</artifactId>
+      <artifactId>kafka-clients</artifactId>
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-integration-test/pom.xml]: The `kafka_2.12` dependency is needed for the Kafka testing infrastructure.  That is why it's only used here in `metron-integration-test`.

File: metron-platform/metron-integration-test/pom.xml [hunk 3]
```diff
@@ -145,6 +145,20 @@
...
+      <groupId>org.apache.kafka</groupId>
+      <artifactId>kafka_2.12</artifactId>
+      <version>${global_kafka_version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.kafka</groupId>
+      <artifactId>kafka_2.12</artifactId>
+      <version>${global_kafka_version}</version>
+      <classifier>test</classifier>
+      <!--scope>test</scope-->
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-integration-test/pom.xml]: This is how Kafka organizes it.

File: metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java [hunk 4]
```diff
@@ -222,8 +243,6 @@ public void stop() {
       zkClient.deleteRecursive(ZkUtils.PreferredReplicaLeaderElectionPath());
       zkClient.deleteRecursive(ZkUtils.BrokerSequenceIdPath());
       zkClient.deleteRecursive(ZkUtils.IsrChangeNotificationPath());
-      zkClient.deleteRecursive(ZkUtils.EntityConfigPath());
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java]: This method no longer exists in the newer version.  I don't know what it's for but test still pass so I assume it isn't needed.

File: metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java [hunk 5]
```diff
@@ -237,43 +256,18 @@ public void reset() {
...
+    }
+    consumer.assign(Arrays.asList(new TopicPartition(topic, 0)));
+    consumer.seek(new TopicPartition(topic, 0), 0);
     List<byte[]> messages = new ArrayList<>();
-    while(results.hasNext()) {
-      ByteBuffer payload = results.next().message().payload();
-      byte[] bytes = new byte[payload.limit()];
-      payload.get(bytes);
-      messages.add(bytes);
+    ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofSeconds(1));
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java]: I will make this a constant.

File: metron-platform/metron-solr/metron-solr-storm/pom.xml [hunk 6]
```diff
@@ -30,30 +30,6 @@
         <commons.config.version>1.10</commons.config.version>
     </properties>
     <dependencies>
-        <dependency>
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-solr/metron-solr-storm/pom.xml]: Correct, these changes were necessary to fix classpath issues.  Newer dependency versions introduce new transitive dependency versions.  Any classpath ordering changes in this PR were necessary.

File: metron-platform/metron-writer/metron-writer-common/src/test/java/org/apache/metron/writer/kafka/KafkaWriterTest.java [hunk 7]
```diff
@@ -93,7 +93,7 @@ public void testHappyPathGlobalConfig() throws Exception {
     assertEquals(producerConfigs.get("bootstrap.servers"), "localhost:6667");
     assertEquals(producerConfigs.get("key.serializer"), "org.apache.kafka.common.serialization.StringSerializer");
     assertEquals(producerConfigs.get("value.serializer"), "org.apache.kafka.common.serialization.StringSerializer");
-    assertEquals(producerConfigs.get("request.required.acks"), 1);
+    assertEquals(producerConfigs.get("acks"), "1");
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-writer/metron-writer-common/src/test/java/org/apache/metron/writer/kafka/KafkaWriterTest.java]: I believe this property changed from `request.required.acks` to `acks` from version 0.9.x to 0.10.x:

0.9.x - https://kafka.apache.org/090/documentation.html#design_ha
2.0.x - https://kafka.apache.org/20/documentation.html#design_ha

File: metron-deployment/ansible/roles/metron-builder/tasks/clean.yml [hunk 8]
```diff
@@ -23,7 +23,7 @@
   args:
     chdir: "{{ metron_build_dir }}"
   with_items:
-    - mvn clean -P HDP-2.5.0.0
+    - mvn clean -P HDP-3.1.0.0
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-deployment/ansible/roles/metron-builder/tasks/clean.yml]: I will revert this back to HDP 2.5 and only upgrade the Kafka version

File: metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java [hunk 9]
```diff
@@ -192,7 +214,6 @@ public String getZookeeperConnect() {
 
   @Override
   public void stop() {
-    shutdownConsumer();
     shutdownProducers();
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-integration-test/src/main/java/org/apache/metron/integration/components/KafkaComponent.java]: Not sure why I originally did this.  I will add it back.

File: metron-platform/metron-pcap-backend/pom.xml [hunk 10]
```diff
@@ -129,21 +129,6 @@
                 </exclusion>
             </exclusions>
         </dependency>
-        <dependency>
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-platform/metron-pcap-backend/pom.xml]: This has been replaced by `kafka-clients`.  This is the only Kafka dependency we need to depend on with the exception of the in memory testing stuff.

File: pom.xml (hunk 1 above)
Comment (merrimanr) [reply pom.xml]: Sure.  I don't believe any changes were needed for Storm.

**Pull request closed without merging**